C7Engine/AI/Pathing/DijkstrasLandAlgorithm.cs

```diff
@@ -0,0 +1,87 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using C7GameData;
+
+namespace C7Engine.Pathing
+{
+	/**
+	 * Uses Dijkstra's Algorithm to find a path between two tiles.
+	 * Advantages: Finds the shortest path accounting for tile movement cost
+	 * Disadvantages: Expensive, will exhaustively search every tile on the continent
+	 *
+	 * Notes: - Edge weight is defined by the movement cost of the 2nd node:
+	 *          w(u, v) = movement_cost(v)
+	 */
+	public class DijkstrasLandAlgorithm : PathingAlgorithm
+	{
+		//N.B. This should really be static, but we can't put a static method on interfaces, so it isn't.
+		public TilePath PathFrom(Tile start, Tile destination)
+		{
+			// shortest distance from start to each tile on the continent
+			Dictionary<Tile, int> dist = new Dictionary<Tile, int>();
+			Dictionary<Tile, Tile> predecessors = new Dictionary<Tile, Tile>();
+			HashSet<Tile> visited = new HashSet<Tile>();
+
+			dist[start] = 0;
+
+			bool updateShortestDistance(Tile tile, int distance) {
+				if (!dist.ContainsKey(tile)) {
+					dist[tile] = distance;
+					return true;
+				} else if (dist[tile] > distance) {
+					dist[tile] = distance;
+					return true;
+				}
+				return false;
+			}
+
+			// TODO not O(n) search
+			KeyValuePair<Tile, int> getNextClosest() {
+				KeyValuePair<Tile, int> next = new KeyValuePair<Tile, int>(Tile.NONE, int.MaxValue);
+				foreach (KeyValuePair<Tile, int> pair in dist) {
+					if (pair.Value < next.Value && !visited.Contains(pair.Key)) {
+						next = pair;
+					}
+				}
+				visited.Add(next.Key);
+				return next;
+			}
+
+			KeyValuePair<Tile, int> closest = getNextClosest();
+			while (closest.Key != Tile.NONE) {
+				foreach (Tile tile in closest.Key.GetLandNeighbors()) {
+					if (!visited.Contains(tile) && updateShortestDistance(tile, closest.Value + tile.MovementCost())) {
+						predecessors[tile] = closest.Key;
+					}
+				}
+				closest = getNextClosest();
+				// TODO: this is fastest if recomputing Dijkstra's for each unit
+				// and ignoring that units may path from the same starting tile
+				if (closest.Key == destination) {
```

Comment:
So, yeah.  This is one of the great questions - should we return early as you have here, or not?

In order to be efficient, we probably do want to have some sort of caching.  But as you point out, it's also not efficient to calculate everything if the situations - edge lengths, in math terms - are going to change before the next unit requests a path, which could happen if a worker builds a road, for example.

Long term, I think caching results will be helpful.  But we'll also have to deal with cache invalidation, one of the three hard problems in computer science.  This is another area where I expect a lot of iteration as we seek to get the optimal results.

For now though, this is perfectly fine.  We need Dijkstra's to work for roads in particular, not to be optimal on a Huge map with 7000 units.

Reply:
I was doing some reading about caching in graph traversal and found a couple of interesting papers on how/when to cache. I only skimmed through them, but it's possible that those concepts may be applicable in our case as well so I'll try and find them to share. I agree we will probably need caching, which will definitely introduce complexity over time